clientv3/maintenance.go

```diff
@@ -53,12 +54,28 @@ type Maintenance interface {
 }
 
 type maintenance struct {
-	c      *Client
-	remote pb.MaintenanceClient
+	dial func(conn *grpc.ClientConn, endpoint string) (pb.MaintenanceClient, error)
 }
 
 func NewMaintenance(c *Client) Maintenance {
-	return &maintenance{c: c, remote: pb.NewMaintenanceClient(c.conn)}
+	return &maintenance{
+		// if endpoint given, update 'conn' with new connection
+		// make sure not to overwrite given *Client.conn from 'c'
+		dial: func(conn *grpc.ClientConn, endpoint string) (pb.MaintenanceClient, error) {
```

Comment:
dial shouldn't have a connection, since it's trying to establish a connection in the first place. Try:
```go
dial: func(endpoint string) (pb.MaintenanceClient, func(), error) {
    conn, err := c.dial(endpoint)
    if err != nil {
        return nil, err
    }
    cancel := func() { conn.Close() }
    return pb.NewMaintenanceClient(conn), cancel, nil
}
```